src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsStreamIterator.java

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright (C) 2015 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.genomics.utils.grpc;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import com.google.api.client.util.BackOff;
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.cloud.genomics.utils.GenomicsFactory;
+import com.google.cloud.genomics.utils.ShardBoundary;
+import com.google.common.base.Predicate;
+import com.google.common.collect.ForwardingIterator;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+
+/**
+ * An iterator for streaming genomic data via gRPC with shard boundary semantics.
+ * 
+ * Includes complex retry logic to upon failure resume the stream at the last known good start
+ * position without returning duplicate data.
+ * 
+ * @param <A> Streaming request type.
+ * @param <B> Streaming response type.
+ * @param <C> Genomic data type returned by stream.
+ * @param <D> Blocking stub type.
+ */
+public abstract class GenomicsStreamIterator<A, B, C, D extends io.grpc.stub.AbstractStub<D>> extends ForwardingIterator<B> {
```

Comment:
Why extend ForwardingIterator? Since you override both hasNext() and next() and you call those directly on the delegate, i.e. not calling super.next() or super.hasNext().